### PR TITLE
Minlibc add asctime and div

### PR DESCRIPTION
**Pull request by sun681**

### CODE REVIEW

**Comment by sun681** [comment]
I am aiming to add much more support in the near future to the posix/c_lib part.
This small pull request is just to see that I'm on track.

File: lib/libc/minimal/source/time/asctime.c [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <time.h>
+#include <bits/restrict.h>
+
+static char *z_impl_asctime(const struct tm *timeptr, char *buf, size_t buflen)
+{
+	static char wday_name[7][3] = {
```
Comment (pabigot) [inline lib/libc/minimal/source/time/asctime.c:13]: `static const char* wday_name[] = {`

The string values should be in ROM.  Then get rid of the precision argument in the sprintf format string.  Or leave it, and switch this to:

    static const char* wdays = "SunMonTueWedThuFriSat???";
    size_t wdidx = ((unsigned int)timeptr->tm_wday < 7U) ? (3U * timeptr->tm_wday) : (3U * 7U);

and use `&wdays[wdidx]`.  See below for range checking.

Whichever you do, do the same for month names.

File: lib/libc/minimal/source/time/asctime.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	return buf;
+}
+
+/*
+ * same as asctime but put it into user supplied buffer.
+ * according to standard the buffer should be at least 26 bytes.
+ */
+char *asctime_r(const struct tm *timeptr, char *buf)
+{
+	return z_impl_asctime(timeptr, buf, MINLIBC_ASCTIME_MIN_BUF_SIZE_REQUIRED);
```
Comment (pabigot) [inline lib/libc/minimal/source/time/asctime.c:43]: Don't use `z_impl_asctime` separately, implement it here.

File: lib/libc/minimal/source/time/asctime.c [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		"Jan", "Feb", "Mar", "Apr", "May", "Jun",
+		"Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
+	};
+
+	if (timeptr == NULL) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	sprintf(buf, "%.3s %.3s%3d %.2d:%.2d:%.2d %d\n",
```
Comment (pabigot) [inline lib/libc/minimal/source/time/asctime.c:26]: Use `snprintf`.

File: lib/libc/minimal/source/time/asctime.c [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+char *asctime_r(const struct tm *timeptr, char *buf)
+{
+	return z_impl_asctime(timeptr, buf, MINLIBC_ASCTIME_MIN_BUF_SIZE_REQUIRED);
+}
+
+char *asctime(const struct tm *timeptr)
+{
+	static char result[MINLIBC_ASCTIME_MIN_BUF_SIZE_REQUIRED];
+
+	return z_impl_asctime(timeptr, result, MINLIBC_ASCTIME_MIN_BUF_SIZE_REQUIRED);
```
Comment (pabigot) [inline lib/libc/minimal/source/time/asctime.c:50]: Use `asctime_r(timeptr, result, sizeof(result))`.

File: lib/libc/minimal/source/time/asctime.c [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	};
+	static char mon_name[12][3] = {
+		"Jan", "Feb", "Mar", "Apr", "May", "Jun",
+		"Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
+	};
+
+	if (timeptr == NULL) {
+		errno = EINVAL;
+		return NULL;
+	}
```
Comment (pabigot) [inline lib/libc/minimal/source/time/asctime.c:24]: Also at minimum validate that all the values you're using as indexes have legal ranges.  You can either reject in that case, or emit `???`, but do not index an invalid pointer.  GNU libc doesn't seem to validate other fields.

File: .checkpatch.conf [hunk 6]
```diff
@@ -24,4 +24,5 @@
 --ignore UNDOCUMENTED_DT_STRING
 --ignore DT_SPLIT_BINDING_PATCH
 --ignore DT_SCHEMA_BINDING_PATCH
+--ignore NEW_TYPEDEFS
```
Comment (pabigot) [inline .checkpatch.conf:27]: No.  This doesn't belong in this PR.

File: lib/libc/minimal/source/time/asctime.c [hunk 7]
```diff
@@ -7,7 +7,6 @@
 #include <stdio.h>
 #include <errno.h>
 #include <time.h>
-#include <bits/restrict.h>
```
Comment (pabigot) [inline lib/libc/minimal/source/time/asctime.c]: Doesn't belong in this PR.

File: lib/libc/minimal/include/time.h [hunk 8]
```diff
@@ -57,11 +57,11 @@ struct tm *gmtime_r(const time_t *_MLIBC_RESTRICT timep,
 /*
  * The asctime/_r functions shall convert the broken-down time in the
  * structure pointed to by timeptr into a string in the form:
- * 
+ *
  *            Day Mon dd hh:mm:ss yyyy\n\0
  */
 char *asctime(const struct tm *timeptr);
-char *__asctime_r (const struct tm *timeptr, char *buf);
+char *asctime_r(const struct tm *timeptr, char *buf);
```
Comment (pabigot) [inline lib/libc/minimal/include/time.h:64]: Squash this into the first commit.

Do **not** open a new PR for this.  If you need help with using git, ask.

**Review by pabigot — CHANGES_REQUESTED** [review]
There's something of a limit on how much we want to duplicate material that can be had by using a real libc, but this seems within that limit.  Some changes needed, though.

File: .checkpatch.conf (hunk 6 above)
Comment (sun681) [reply .checkpatch.conf:27]: Without this line I can not add the div_t and ldiv_t as typeddef structs because of the ci checks.
as far as I saw it should be typedefs according to the standard and use examples i saw online.
Am I right? if yes, you want me to open another pr only for this?

Comment (pabigot) [reply .checkpatch.conf:27]: No.  In this case the CI checks are wrong and must be ignored.

Comment (sun681) [reply .checkpatch.conf:27]: How can I ignore it?

Comment (pabigot) [reply .checkpatch.conf:27]: Accept that that part of the compliance check will fail.  It's the responsibility of the reviewers to understand that's not your problem to solve.

File: lib/libc/minimal/include/time.h (hunk 8 above)
Comment (sun681) [reply lib/libc/minimal/include/time.h:64]: Can you please guide me here? I finished the changes for now and want to make sure I will do this part good

Comment (pabigot) [reply lib/libc/minimal/include/time.h:64]: OK, here are some instructions based on what's currently pushed, and assuming you're using a commandline git interface (if you're using something else, I can't help you).

<details><summary>Expand for some notes</summary>

Here are your commits.
```
* 0655567cd9 (HEAD -> pr/31384) [Mon Jan 18 01:03:55 2021 +0200] lib/libc/minimal: add ignore flag for new typedefs in code
* 920f28f0b6 [Mon Jan 18 00:14:25 2021 +0200] lib/libc/minimal: coding-style errors and fix asctime_r function name
* c70feceb1e [Sun Jan 17 23:33:42 2021 +0200] first commit: add div and asctime impl
```

I want the following:

First, remove the top commit.  First, the description is wrong, as it affects commits throughout the tree, not just ones in lib/libc/minimal.  Second, it overrides a policy decision, so has to be handled separately.  It can be removed by discarding the commit entirely by resetting the head and the workspace to the previous commit.

    git reset --hard HEAD^

Now you have these commits:

```
* 920f28f0b6 (HEAD -> pr/31384) [Mon Jan 18 00:14:25 2021 +0200] lib/libc/minimal: coding-style errors and fix asctime_r function name
* c70feceb1e [Sun Jan 17 23:33:42 2021 +0200] first commit: add div and asctime impl
```

The top commit here has two changes, the removal of the include which I don't want done, and two changes to the `time.h` header that need to be done in the first commit.

First, remove the commit but leave the changes in your workspace:

    git reset --soft HEAD^
    
```
tirzah[1589]$ git status -uno
On branch pr/31384
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   lib/libc/minimal/include/time.h
	modified:   lib/libc/minimal/source/time/asctime.c

Untracked files not listed (use -u option to show untracked files)
```

Then take those out of the staging area:

    git reset

```
tirzah[1592]$ git status -uno
On branch pr/31384
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   lib/libc/minimal/include/time.h
	modified:   lib/libc/minimal/source/time/asctime.c
```

Verify that the asctime.c changes are just the include file:

    git diff lib/libc/minimal/source/time/asctime.c

then discard them:

    git checkout lib/libc/minimal/source/time/asctime.c

Now the only changes are to time.h.  Use diff on the workspace to see the change to the whitespace in the commits, and the renamed function declaration:

    git diff

Now amend the first commit so all the changes are in it:

    git commit --amend -C HEAD -a

Now you have one commit with all your changes.

Unfortunately, having done this, I can tell there's more to do.  The addition of the div support and its test, and the addition of the asctime implementation and its test, also need to be in separate commits.  Also the define for the length of the asctime buffer does not belong in the `time.h` header: it's not part of the C standard library.  So you need to pick up the pieces and commit them separately.

For that you'd want to run:

    git reset --soft HEAD^
    git reset
    git add -i

to get the changes into your workspace without staging them, and then incrementally add pieces.

Then the commit messages need to be cleaned up a lot.
</details>

I'm sorry, but there's a lot to fix here, and describing all the steps is difficult.  I went through the steps above, then a lot of other steps, and ended up with https://github.com/pabigot/zephyr/commits/pr/31384 which is close to what I would like this to look like: two commits, one for each feature with test, fixes incorporated.

I'd suggest experimenting in your workspace to see if you can make what you have look like that.  Look at the documentation for git reset and add and other subcommands for more help.

Comment (sun681) [reply lib/libc/minimal/include/time.h:64]: Hi, im about to finish everything you guided me to do.
but I have 2 questions, first, why you don't want me to remove the include to restrict.h file from the asctime.c? I added this by mistake anyway..
The second questions is what is the right way to contribute to this project without squash my commit all the time, I mean, if im going to do it in a big contribute im about to lose the option to go back in history because all the commits just squash together, no?

Comment (pabigot) [reply lib/libc/minimal/include/time.h:64]: > but I have 2 questions, first, why you don't want me to remove the include to restrict.h file from the asctime.c? I added this by mistake anyway..

I mis-read something, and thought this was something that was present before your patch. If you added it by mistake, then yes remove it.

> The second questions is what is the right way to contribute to this project without squash my commit all the time, I mean, if im going to do it in a big contribute im about to lose the option to go back in history because all the commits just squash together, no?

There's some explanation in the [contribution documentation](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow), but basically we want each commit to be close to self-contained as possible while only addressing one item at a time.

The version you have now:
* adds both asctime and div in a single commit.  They're different  features, so they should be separate.  Some reviewers want the tests in a separate commit; unless they're large and new I personally prefer them in the commit that provides/changes whatever the test uses.  I don't think we have a clear policy on this, but I maybe just can't find it.
* has a second commit that fixes problems in the first one.  This confuses and annoys reviewers who look at the changes one commit at a time, notice the problem and write a comment about it, then go on to the next commit where it's fixed making the comment irrelevant or incorrect.
* has a third commit that is a policy change that doesn't belong in a PR that is adding functionality to minimal libc.

In short we want to be able to look at each commit and verify that it is complete and does what it's supposed to.  So that's the sort of PR you should supply.    When the code is merged we don't need to see your development history, just the final accepted version.  If you want to track history you can keep branches or tags for earlier revisions in your local development workspace.

Comment (sun681) [reply lib/libc/minimal/include/time.h:62]: Thank you for your answers,I work in a place where we always look only in the last commit of a feature branch to see if we can merge it. (of course each branch is only for one feature which is not the case in my original pull request and its my fault).

If for example i would only have added the div feature, and lets say i would do couple of commit, why would you bother yourself to look at each of the commits separately? why not just to review the last one?

Comment (sun681) [reply lib/libc/minimal/include/time.h:62]: and we merge the code to the master with squah of course.

File: tests/lib/c_lib/src/main.c [hunk 9]
```diff
@@ -618,6 +618,15 @@ void test_asctime(void)
 	zassert_true(strcmp(asctime(&time), "??? ???  3 02:01:00 1905\n") == 0, "asctime error");
 }
 
+void test_div(void)
+{
+	/* test example from man */
+	div_t q = div(-5, 3);
```
Comment (pabigot) [inline tests/lib/c_lib/src/main.c:624]: Please extend this to cover all combinations of positive and negative arguments, and for each one verify that the result invariant is correct.  Something like:
```
struct input_s {
   int num;
   int den;
} const input[] = {
   { 5, 3 },
   { 5, 7 },
   { -5, 7 },
   { 5, -7 },
   { -5, -7 },
};

for (size_t i = 0; i < ARRAY_SIZE(input); ++i) {
   int num = input[i].num;
   int den = input[i].den;
   
   div_t res = div(num, den);
   zassert_equal(num, res.quot * den + rem,
                 "Invariant failed for %d / %d : q = %d, r = %d",
		 num, den, res.quot, res.rem);
   zassert_equal(num >= 0, res.rem >= 0,
                 "Quotient not rounded toward zero : q = %d, r = %d",
		 num, den, res.quot, res.rem);
}
```
Also provide a similar test for `ldiv`.

File: tests/lib/c_lib/src/main.c [hunk 10]
```diff
@@ -598,6 +599,34 @@ void test_strtok_r(void)
...
+		.tm_mon = 4,
+		.tm_year = 5,
+		.tm_wday = 6
+	};
+
+	zassert_true(strcmp(asctime(&time), "Sat May  3 02:01:00 1905\n") == 0, "asctime error");
+	time.tm_wday = -2;
+	zassert_true(strcmp(asctime(&time), "??? May  3 02:01:00 1905\n") == 0, "asctime error");
+	time.tm_mon = 15;
+	zassert_true(strcmp(asctime(&time), "??? ???  3 02:01:00 1905\n") == 0, "asctime error");
```
Comment (pabigot) [inline tests/lib/c_lib/src/main.c:618]: Please add a test for `asctime_r` as well.

**Review by pabigot — CHANGES_REQUESTED** [review]
The commits need to include a body that describes what they do; the summary line isn't enough.  For example:

Add support for standard functions to convert encoded calendar time to a text string.

and

Add support for standard API to produce both the quotient and remainder from integer division.

The tests also need to be more complete.

File: lib/libc/minimal/include/time.h (hunk 8 above)
Comment (pabigot) [reply lib/libc/minimal/include/time.h:62]: > why not just to review the last one?

Because we *don't* squash the PR when merged.  Some PRs require as many as 40 commits to carefully build to the completion of the feature they add.  With changes of that size it's critical that each one be as self-contained and as small as possible, but no smaller.

File: tests/lib/c_lib/src/main.c (hunk 9 above)
Comment (sun681) [reply tests/lib/c_lib/src/main.c:624]: about the commit messages, I'll do it.
about adding more tests, i saw this at the top of the file:

` IMPORTANT: The module only ensures that each supported library is present,
 * and that a bare minimum of its functionality is operating correctly. It does
 * NOT guarantee that ALL standards-defined functionality is present, nor does
 * it guarantee that ALL functionality provided is working correctly.`

my guess was that they try to minimize tests for simplicity of the file and they allowed it beacuse the functions logics which this file tests is pretty simple. 
of course I would add it if you still want me to.

Comment (pabigot) [reply tests/lib/c_lib/src/main.c:624]: I don't think that's justifying minimizing tests for provided functions; it's explaining that not all functions are supported.  Please do provide full testing for any function added.

**Review by pabigot — CHANGES_REQUESTED** [review]
Need to reduce the line lengths in the commit messages, and address the compliance check failures except for the one about adding a new typedef.

For the `LINE_SPACING` error if you move the `=` from the wrapped line of `static const char *months` to the end of that line the test might pass.

Use `scripts/ci/check_compliance.py -c HEAD~2..` to test locally.  (That checks compliance only on the last two commits, which are yours.)

File: tests/lib/c_lib/src/main.c [hunk 11]
```diff
@@ -644,6 +644,68 @@ void test_asctime_r(void)
...
+	};
+
+	for (size_t i = 0; i < ARRAY_SIZE(input); ++i) {
+		int num = input[i].num;
+		int den = input[i].den;
+
+		div_t res = div(num, den);
+
+		zassert_equal(num, res.quot * den + res.rem,
+						"Invariant failed for %d / %d : q = %d, r = %d",
```
Comment (pabigot) [inline tests/lib/c_lib/src/main.c:669]: The argument indentation is still inconsistent, here and in `test_ldiv`.

File: tests/lib/c_lib/src/main.c [hunk 12]
```diff
@@ -598,6 +599,113 @@ void test_strtok_r(void)
...
+	};
+
+	for (size_t i = 0; i < ARRAY_SIZE(input); ++i) {
+		int num = input[i].num;
+		int den = input[i].den;
+
+		div_t res = div(num, den);
+
+		zassert_equal(num, res.quot * den + res.rem,
+					"Invariant failed for %d / %d : q = %d, r = %d",
```
Comment (pabigot) [inline tests/lib/c_lib/src/main.c:669]: This is still not right.  Tabs in Zephyr are eight spaces; it seems you have tabs at four spaces.  When viewed here on github you want it to look like:
```
                zassert_equal(num, res.quot * den + res.rem,
                        "Invariant failed for %ld / %ld : q = %ld, r = %ld",
                        num, den, res.quot, res.rem);
```
with 3 tabs and no spaces on the wrapped lines, or
```
                zassert_equal(num, res.quot * den + res.rem,
                              "Invariant failed for %ld / %ld : q = %ld, r = %ld",
                              num, den, res.quot, res.rem);
```
by adding spaces after the last tab to align with the first line.

**Review by pabigot — CHANGES_REQUESTED** [review]

**Review by pabigot — APPROVED** [review]
Thanks.  Note that we're in stabilization phase right now, so this may not get merged until the 2.6 merge window opens.

To other reviewers: the compliance checks are a false positive: this *must* provide the standard-defined typedefs.
